metron-platform/metron-writer/src/main/java/org/apache/metron/writer/BulkWriterComponent.java

```diff
@@ -18,315 +18,159 @@
- *   batchTimeout, and also less than the 'topology.message.timeout.secs'.  This guarantees that the messages
- *   will not age out of the Storm topology, but it does not guarantee the flush interval requested, for
- *   sensor types not receiving at least one message every second.
+ * This component manages an internal cache of messages to be written in batch.  A separate cache is used for each sensor.
+ * Each time a message is written to this component, flush policies are applied to determine if a batch of messages should
+ * be flushed.  When a flush does happen, this component calls the supplied {@link org.apache.metron.writer.BulkWriterResponseHandler}
+ * so that the calling class can handle any necessary message acknowledgement.  This component also ensures all messages
+ * in a batch are included in the response as either a success or failure.
  *
  * @param <MESSAGE_T>
  */
 public class BulkWriterComponent<MESSAGE_T> {
-  public static final Logger LOG = LoggerFactory
-            .getLogger(BulkWriterComponent.class);
-  private Map<String, Collection<Tuple>> sensorTupleMap = new HashMap<>();
-  private Map<String, List<MESSAGE_T>> sensorMessageMap = new HashMap<>();
-  private Map<String, long[]> batchTimeoutMap = new HashMap<>();
-  private OutputCollector collector;
-  //In test scenarios, defaultBatchTimeout may not be correctly initialized, so do it here.
-  //This is a conservative defaultBatchTimeout for a vanilla bolt with batchTimeoutDivisor=2
-  public static final int UNINITIALIZED_DEFAULT_BATCH_TIMEOUT = 6;
-  private int defaultBatchTimeout = UNINITIALIZED_DEFAULT_BATCH_TIMEOUT;
-  private boolean handleCommit = true;
-  private boolean handleError = true;
-  private static final int LAST_CREATE_TIME_MS = 0; //index zero'th element of long[] in batchTimeoutMap
-  private static final int TIMEOUT_MS = 1;          //index next element of long[] in batchTimeoutMap
-  private Clock clock = new Clock();
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private Map<String, List<BulkWriterMessage<MESSAGE_T>>> sensorMessageCache = new HashMap<>();
+  private BulkWriterResponseHandler bulkWriterResponseHandler;
+  private BatchTimeoutPolicy batchTimeoutPolicy;
+  private List<FlushPolicy> flushPolicies;
 
-  public BulkWriterComponent(OutputCollector collector) {
-    this.collector = collector;
-  }
-
-  public BulkWriterComponent(OutputCollector collector, boolean handleCommit, boolean handleError) {
-    this(collector);
-    this.handleCommit = handleCommit;
-    this.handleError = handleError;
+  public BulkWriterComponent(BulkWriterResponseHandler bulkWriterResponseHandler) {
+    this.bulkWriterResponseHandler = bulkWriterResponseHandler;
+    flushPolicies = new ArrayList<>();
+    flushPolicies.add(new BatchSizePolicy());
+    batchTimeoutPolicy = new BatchTimeoutPolicy();
```

Comment:
Small nit - instead of having a special instance variable for batchTimeoutPolicy that later has a special dep injection method for clock, could we instead add a protected constructor that allows you to pass in the default flush policies, or maybe just the clock? When I see the constructor that has special handling for one of the policies as a private member variable, it leads me to wondering why it's treated special. I don't have a strong preference for either approach, though passing in the clock is a bit more clear that we're testing with default policies and only injecting variation with the provided clock. 

In general, I have a slight preference for constructors over setter methods for dep injection because the options for test injections are very clearly documented by the additional constructors vs having to look for n setter methods in the API. It also forces the object to be fully configured with an expected valid state _prior_ to calling any methods under test, which I think is easier for future developers making changes to the same classes and tests.

Here are a couple examples:

1. 
```
// Using clock as constructor arg
/**
 * Main constructor
 **/
public BulkWriterComponent(BulkWriterResponseHandler bulkWriterResponseHandler) {
    this(bulkWriterResponseHandler, new Clock());
}

public BulkWriterComponent(BulkWriterResponseHandler bulkWriterResponseHandler, Clock clock) {
    this.bulkWriterResponseHandler = bulkWriterResponseHandler;
    flushPolicies = new ArrayList<>();
    flushPolicies.add(new BatchSizePolicy());
    flushPolicies.add(new BatchTimeoutPolicy(clock));
}
```
2.
```
// Using flush polilcies as constructor arg
/**
 * Main constructor
 **/
public BulkWriterComponent(BulkWriterResponseHandler bulkWriterResponseHandler) {
    List<FlushPolicy> flushPolicies = new ArrayList<>();
    flushPolicies.add(new BatchSizePolicy());
    flushPolicies.add(new BatchTimeoutPolicy(new Clock()));
    this(bulkWriterResponseHandler, flushPolicies);
}

public BulkWriterComponent(BulkWriterResponseHandler bulkWriterResponseHandler, List<FlushPolicy> flushPolicies) {
    this.bulkWriterResponseHandler = bulkWriterResponseHandler;
    this.flushPolicies = flushPolicies;
}
```

Reply:
The reason I did this is because of `batchTimeoutPolicy`.  There is a `BulkWriterComponent.setDefaultBatchTimeout` method and it's not obvious it's only used for testing (like clock is).  I didn't want to assume this can only be set at `BulkWriterComponent` creation time.  It looks like it's only set in the bolt prepare methods so I believe we can safely move it to the constructor like you're suggesting.  Do we want to make this assumption, that `batchTimeoutPolicy` is only set once?  I will make the clock change either way if you think it makes the code easier to follow.

Reply:
I see what you're saying. Honestly, I think the batch timeout value belongs to the policy, not the component. I get providing a default via the component, but BulkWriterComponent.setDefaultBatchTimeout implies a hard-coded dependency on batch timeout policy, whereas at the same time we're trying to abstract the component away from specific policy implementations. I think we should either provide the timeout externally as an arg (like my other recommendations), or provide config that's pluggable for the policies. E.g. if we need to update that timeout on the fly (I think we do, actually), then have the class that provided the policies in the first place update the config for that policy. I think that might be in the bolt, and this sounds like it's related to the config supplier stuff we do with Curator. 

Another idea for managing default policies could be to put the construction of that list into a default factory of sorts. Just a thought, I'm ok either way.

Reply:
@merrimanr - Thought about this a bit more in context of @nickwallen's recent comments. Yeah, while the actual timeout setting itself can change, the _default_ should definitely not change on the fly. I can't think of a good reason why we would ever want a default to dynamically change. That's more of an upgrade/new release type of change imo that has documentation and release note implications.

Reply:
Thanks that helps.  I will play around with this and see if I can come up with a solution based on your and @nickwallen's feedback.